File: app/src/main/java/com/annushkaproject/programmerscalculator/utils/ProgrammerOperationsUtil.java
package com.annushkaproject.programmerscalculator.utils;

import com.annushkaproject.programmerscalculator.model.ProgrammerCalcModel;
import com.annushkaproject.programmerscalculator.model.int_size_enum;

import java.math.BigDecimal;

public class ProgrammerOperationsUtil {
    /**
     * Used to get calculated result.
     * @param data Model used to calculate the result.
     * @return Result.
     */
    public static long calculateWithData(ProgrammerCalcModel data) {
        switch (data.getOperator()) {
            case ADD:
                return addSubtract(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case SUBTRACT:
                return addSubtract(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue().negate());
            case MULTIPLY:
                return multiply(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case remainder_divide:
                return divide(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case LSH:
                return lsh(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case RSH:
                return rsh(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case OR:
                return or(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case XOR:
                return xor(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case AND:
                return and(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case MOD:
                return mod(data.getBytelengthenum(), data.getFirstValue(), data.getSecondValue());
            case CHANGE_SIGN:
                return changeSign(data.getBytelengthenum(), data.getFirstValue());
            case NOT:
                return not(data.getBytelengthenum(), data.getFirstValue());
            default:
                return 0;
        }
    }

    private static long addSubtract(int_size_enum mode, BigDecimal value1, BigDecimal value2) {
        switch (mode) {
            case int_size_enum_kvrd:
                return value1.longValue() + value2.longValue();
            case int_size_enum_dvrd:
                return value1.intValue() + value2.intValue();
            case int_size_enum_vrd:
                return (short) (value1.shortValue() + value2.shortValue());
            case int_size_enum_byte:
                return (byte) (value1.byteValue() + value2.byteValue());
            default:
                return 0;
        }
    }

    private static long changeSign(int_size_enum mode, BigDecimal value) {
        switch (mode) {
            case int_size_enum_kvrd:
                return value.negate().longValue();
            case int_size_enum_dvrd:
                return value.negate().intValue();
            case int_size_enum_vrd:
                return value.negate().shortValue();
            case int_size_enum_byte:
                return value.negate().byteValue();
            default:
                return 0;
        }
    }

    private static long lsh(int_size_enum mode, BigDecimal value, BigDecimal shift) {
        switch (mode) {
            case int_size_enum_kvrd:
                return value.longValue() << shift.intValue();
            case int_size_enum_dvrd:
                return value.intValue() << shift.intValue();
            case int_size_enum_vrd:
                return (short) (value.shortValue() << shift.intValue());
            case int_size_enum_byte:
                return (byte) (value.byteValue() << shift.intValue());
            default:
                return 0;
        }
    }

    private static long rsh(int_size_enum mode, BigDecimal value, BigDecimal shift) {
        switch (mode) {
            case int_size_enum_kvrd:
                return value.longValue() >> shift.intValue();
            case int_size_enum_dvrd:
                return value.intValue() >> shift.intValue();
            case int_size_enum_vrd:
                return (short) (value.shortValue() >> shift.intValue());
            case int_size_enum_byte:
                return (byte) (value.byteValue() >> shift.intValue());
            default:
                return 0;
        }
    }

    private static long not(int_size_enum mode, BigDecimal value) {
        switch (mode) {
            case int_size_enum_kvrd:
                return ~value.longValue();
            case int_size_enum_dvrd:
                return ~value.intValue();
            case int_size_enum_vrd:
                return (short) (~value.shortValue());
            case int_size_enum_byte:
                return (byte) (~value.byteValue());
            default:
                return 0;
        }
    }

    private static long or(int_size_enum mode, BigDecimal value1, BigDecimal value2) {
        switch (mode) {
            case int_size_enum_kvrd:
                return value1.longValue() | value2.longValue();
            case int_size_enum_dvrd:
                return value1.intValue() | value2.intValue();
            case int_size_enum_vrd:
                return (short) (value1.shortValue() | value2.shortValue());
            case int_size_enum_byte:
                return (byte) (value1.byteValue() | value2.byteValue());
            default:
                return 0;
        }
    }

    private static long xor(int_size_enum mode, BigDecimal value1, BigDecimal value2) {
        switch (mode) {
            case int_size_enum_kvrd:
                return value1.longValue() ^ value2.longValue();
            case int_size_enum_dvrd:
                return value1.intValue() ^ value2.intValue();
            case int_size_enum_vrd:
                return (short) (value1.shortValue() ^ value2.shortValue());
            case int_size_enum_byte:
                return (byte) (value1.byteValue() ^ value2.byteValue());
            default:
                return 0;
        }
    }

    private static long and(int_size_enum mode, BigDecimal value1, BigDecimal value2) {
        switch (mode) {
            case int_size_enum_kvrd:
                return value1.longValue() & value2.longValue();
            case int_size_enum_dvrd:
                return value1.intValue() & value2.intValue();
            case int_size_enum_vrd:
                return (short) (value1.shortValue() & value2.shortValue());
            case int_size_enum_byte:
                return (byte) (value1.byteValue() & value2.byteValue());
            default:
                return 0;
        }
    }

    private static long mod(int_size_enum mode, BigDecimal dividend, BigDecimal divider) {
        switch (mode) {
            case int_size_enum_kvrd:
                return dividend.longValue() % divider.longValue();
            case int_size_enum_dvrd:
                return dividend.intValue() % divider.intValue();
            case int_size_enum_vrd:
                return (short) (dividend.shortValue() % divider.shortValue());
            case int_size_enum_byte:
                return (byte) (dividend.byteValue() % divider.byteValue());
            default:
                return 0;
        }
    }

    private static long multiply(int_size_enum mode, BigDecimal value1, BigDecimal value2) {
        switch (mode) {
            case int_size_enum_kvrd:
                return value1.longValue() * value2.longValue();
            case int_size_enum_dvrd:
                return value1.intValue() * value2.intValue();
            case int_size_enum_vrd:
                return (short) (value1.shortValue() * value2.shortValue());
            case int_size_enum_byte:
                return (byte) (value1.byteValue() * value2.byteValue());
            default:
                return 0;
        }
    }

    private static long divide(int_size_enum mode, BigDecimal dividend, BigDecimal divider) {
        switch (mode) {
            case int_size_enum_kvrd:
                return dividend.longValue() / divider.longValue();
            case int_size_enum_dvrd:
                return dividend.intValue() / divider.intValue();
            case int_size_enum_vrd:
                return (short) (dividend.shortValue() / divider.shortValue());
            case int_size_enum_byte:
                return (byte) (dividend.byteValue() / divider.byteValue());
            default:
                return 0;
        }
    }
}

File: app/src/main/java/com/annushkaproject/programmerscalculator/utils/ProgrammerOperationsUtil.kt
package com.annushkaproject.programmerscalculator.utils

import com.annushkaproject.programmerscalculator.model.Operator
import com.annushkaproject.programmerscalculator.model.ProgrammerCalcModel
import com.annushkaproject.programmerscalculator.model.int_size_enum
import java.math.BigDecimal

object ProgrammerOperationsUtil {
    @JvmStatic
    fun calculateWithData(data: ProgrammerCalcModel): Long {
        return when (data.operator) {
            Operator.ADD -> addSubtract(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.SUBTRACT -> addSubtract(data.bytelengthenum, data.firstValue, data.secondValue.negate())
            Operator.MULTIPLY -> multiply(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.remainder_divide -> divide(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.LSH -> lsh(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.RSH -> rsh(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.OR -> or(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.XOR -> xor(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.AND -> and(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.get_remainder -> mod(data.bytelengthenum, data.firstValue, data.secondValue)
            Operator.CHANGE_SIGN -> changeSign(data.bytelengthenum, data.firstValue)
            Operator.NOT -> not(data.bytelengthenum, data.firstValue)
            else -> 0
        }
    }

    private fun addSubtract(mode: int_size_enum, value1: BigDecimal, value2: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value1.toLong() + value2.toLong()
            int_size_enum.l4 -> (value1.toInt() + value2.toInt()).toLong()
            int_size_enum.l2 -> (value1.toShort() + value2.toShort()).toShort()
            int_size_enum.l1 -> (value1.toByte() + value2.toByte()).toByte()
            else -> 0
        }
    }

    private fun changeSign(mode: int_size_enum, value: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value.negate().toLong()
            int_size_enum.l4 -> value.negate().toInt().toLong()
            int_size_enum.l2 -> value.negate().toShort().toLong()
            int_size_enum.l1 -> value.negate().toByte().toLong()
            else -> 0
        }
    }

    private fun lsh(mode: int_size_enum, value: BigDecimal, shift: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value.toLong() shl shift.toInt()
            int_size_enum.l4 -> (value.toInt() shl shift.toInt()).toLong()
            int_size_enum.l2 -> ((value.toShort() shl shift.toInt()) as Short).toLong()
            int_size_enum.l1 -> ((value.toByte() shl shift.toInt()) as Byte).toLong()
            else -> 0
        }
    }

    private fun rsh(mode: int_size_enum, value: BigDecimal, shift: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value.toLong() shr shift.toInt()
            int_size_enum.l4 -> (value.toInt() shr shift.toInt()).toLong()
            int_size_enum.l2 -> ((value.toShort() shr shift.toInt()) as Short).toLong()
            int_size_enum.l1 -> ((value.toByte() shr shift.toInt()) as Byte).toLong()
            else -> 0
        }
    }

    private fun not(mode: int_size_enum, value: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value.toLong().inv()
            int_size_enum.l4 -> value.toInt().inv().toLong()
            int_size_enum.l2 -> (value.toShort().inv() as Short).toLong()
            int_size_enum.l1 -> (value.toByte().inv() as Byte).toLong()
            else -> 0
        }
    }

    private fun or(mode: int_size_enum, value1: BigDecimal, value2: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value1.toLong() or value2.toLong()
            int_size_enum.l4 -> (value1.toInt() or value2.toInt()).toLong()
            int_size_enum.l2 -> ((value1.toShort() or value2.toShort()) as Short).toLong()
            int_size_enum.l1 -> ((value1.toByte() or value2.toByte()) as Byte).toLong()
            else -> 0
        }
    }

    private fun xor(mode: int_size_enum, value1: BigDecimal, value2: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value1.toLong() xor value2.toLong()
            int_size_enum.l4 -> (value1.toInt() xor value2.toInt()).toLong()
            int_size_enum.l2 -> ((value1.toShort() xor value2.toShort()) as Short).toLong()
            int_size_enum.l1 -> ((value1.toByte() xor value2.toByte()) as Byte).toLong()
            else -> 0
        }
    }

    private fun and(mode: int_size_enum, value1: BigDecimal, value2: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value1.toLong() and value2.toLong()
            int_size_enum.l4 -> (value1.toInt() and value2.toInt()).toLong()
            int_size_enum.l2 -> ((value1.toShort() and value2.toShort()) as Short).toLong()
            int_size_enum.l1 -> ((value1.toByte() and value2.toByte()) as Byte).toLong()
            else -> 0
        }
    }

    private fun mod(mode: int_size_enum, dividend: BigDecimal, divider: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> dividend.toLong() % divider.toLong()
            int_size_enum.l4 -> (dividend.toInt() % divider.toInt()).toLong()
            int_size_enum.l2 -> (dividend.toShort() % divider.toShort()).toShort()
            int_size_enum.l1 -> (dividend.toByte() % divider.toByte()).toByte()
            else -> 0
        }
    }

    private fun multiply(mode: int_size_enum, value1: BigDecimal, value2: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> value1.toLong() * value2.toLong()
            int_size_enum.l4 -> (value1.toInt() * value2.toInt()).toLong()
            int_size_enum.l2 -> (value1.toShort() * value2.toShort()).toShort()
            int_size_enum.l1 -> (value1.toByte() * value2.toByte()).toByte()
            else -> 0
        }
    }

    private fun divide(mode: int_size_enum, dividend: BigDecimal, divider: BigDecimal): Long {
        return when (mode) {
            int_size_enum.l8 -> dividend.toLong() / divider.toLong()
            int_size_enum.l4 -> (dividend.toInt() / divider.toInt()).toLong()
            int_size_enum.l2 -> (dividend.toShort() / divider.toShort()).toShort()
            int_size_enum.l1 -> (dividend.toByte() / divider.toByte()).toByte()
            else -> 0
        }
    }
}